server/index.ts

```diff
@@ -1,13 +1,132 @@
+}
+
+const consumerKey = process.env.DISCOGS_CONSUMER_KEY;
+const consumerSecret = process.env.DISCOGS_CONSUMER_SECRET;
 
 app.get('/api/hello', (_request, response) => {
   response.json({ message: 'Hello from server' });
 });
 
+let oAuthRequestTokenSecret: string | null;
+let oAuthRequestToken: string | null;
+let oAuthAccessTokenSecret: string | null;
+let oAuthAccessToken: string | null;
+
+app.get('/api/oauth/request_token', async (_request, response) => {
+  const tokenResponse = await fetch(
+    'https://api.discogs.com/oauth/request_token',
+    {
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        Authorization: `OAuth oauth_consumer_key="${consumerKey}",oauth_signature_method="PLAINTEXT",oauth_timestamp="${Date.now()}",oauth_nonce="${Date.now()}",oauth_version="1.0",oauth_signature="${consumerSecret}&", oauth_callback="http://localhost:3001/api/oauth/return"`,
+      },
+    }
+  );
+
+  const token = await tokenResponse.text();
+  const params = new URLSearchParams(token);
+  oAuthRequestToken = params.get('oauth_token');
+  oAuthRequestTokenSecret = params.get('oauth_token_secret');
+  response.redirect(
+    `https://discogs.com/oauth/authorize?oauth_token=${oAuthRequestToken}`
+  );
+});
+
+app.get('/api/oauth/return', async (request, response) => {
+  const { oauth_verifier: oAuthVerifier } = request.query;
+
+  const tokenResponse = await fetch(
+    'https://api.discogs.com/oauth/access_token',
+    {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        Authorization: `OAuth oauth_consumer_key="${consumerKey}",oauth_nonce="${Date.now()}",oauth_token="${oAuthRequestToken}", oauth_signature=${consumerSecret}&${oAuthRequestTokenSecret}",oauth_signature_method="PLAINTEXT",oauth_timestamp="${Date.now()}",oauth_verifier="${oAuthVerifier}"`,
+      },
+    }
+  );
+
+  const token = await tokenResponse.text();
+  const params = new URLSearchParams(token);
+  oAuthAccessToken = params.get('oauth_token');
+  oAuthAccessTokenSecret = params.get('oauth_token_secret');
+
+  const identityResponse = await fetch(
+    'https://api.discogs.com/oauth/identity',
+    {
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        Authorization: `OAuth oauth_consumer_key="${consumerKey}", oauth_nonce="${Date.now()}", oauth_token="${oAuthAccessToken}", oauth_signature="${consumerSecret}&${oAuthAccessTokenSecret}",oauth_signature_method="PLAINTEXT",oauth_timestamp="${Date.now()}"`,
+      },
+    }
+  );
+  const identity = await identityResponse.json();
+  const username = identity.username;
+  if (oAuthAccessToken && oAuthAccessTokenSecret) {
+    response.cookie('username', username, { httpOnly: true });
+    response.cookie('accessToken', oAuthAccessToken, { httpOnly: true });
+    response.cookie('accessTokenSecret', oAuthAccessTokenSecret, {
+      httpOnly: true,
+    });
```

Comment:
Instead of three cookies, you could reduce it to a single cookie by creating a JSON.

e.g.

```suggestion
    response.cookie('auth', JSON.stringify({ username, oAuthAccessToken, oAuthAccessTokenSecret }), { httpOnly: true });
```

You need to parse it later.

If you like to encrypt it, please take a look at json-web-token.